x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java

```diff
@@ -232,19 +235,136 @@ private static void validateField(Map<?, ?> properties, String fieldName, boolea
         }
     }
 
-    private XContentBuilder resolveEnrichMapping(final EnrichPolicy policy) {
-        // Currently the only supported policy type is EnrichPolicy.MATCH_TYPE, which is a keyword type
-        final String keyType;
-        final CheckedFunction<XContentBuilder, XContentBuilder, IOException> matchFieldMapping;
+    private void resolveEnrichMapping(
+        final EnrichPolicy policy,
+        GetIndexResponse getIndexResponse,
+        ActionListener<XContentBuilder> resultListener
+    ) {
         if (EnrichPolicy.MATCH_TYPE.equals(policy.getType())) {
-            matchFieldMapping = (builder) -> builder.field("type", "keyword").field("doc_values", false);
-            // No need to also configure index_options, because keyword type defaults to 'docs'.
+            createEnrichMappingBuilder((builder) -> builder.field("type", "keyword").field("doc_values", false), resultListener);
+        } else if (EnrichPolicy.RANGE_TYPE.equals(policy.getType())) {
+            resolveRangeTypeEnrichMapping(policy, getIndexResponse, resultListener);
         } else if (EnrichPolicy.GEO_MATCH_TYPE.equals(policy.getType())) {
-            matchFieldMapping = (builder) -> builder.field("type", "geo_shape");
+            createEnrichMappingBuilder((builder) -> builder.field("type", "geo_shape"), resultListener);
         } else {
-            throw new ElasticsearchException("Unrecognized enrich policy type [{}]", policy.getType());
+            resultListener.onFailure(new ElasticsearchException("Unrecognized enrich policy type [{}]", policy.getType()));
         }
+    }
 
+    private void resolveRangeTypeEnrichMapping(
+        EnrichPolicy policy,
+        GetIndexResponse response,
```

Comment:
Maybe pass down here a `Map<String, Object>` instead of a `GetIndexResponse`? Since that the only thing from the response used here.

Reply:
And then maybe replace `collapseFieldMappingsValue(...)` and `property(...)` method with `ObjectPath#eval(...)`?

Reply:
`ObjectPath#eval(...)` was definitely the thing I was looking for :sweat_smile: 

Passing around the map makes it a bit unclear whats in the map / structure of it, I don't know if that sounds weird?

Reply:
> Passing around the map makes it a bit unclear whats in the map / structure of it, I don't know if that sounds weird?

I think if the parameter is named `mappingSource` or something like that, that it isn't an issue.

Reply:
Addressed in https://github.com/elastic/elasticsearch/pull/76110/commits/44cab8e54cb59099114a07af78626c1130a74d09